Major refactor to add Ubuntu (16.04) support, migrate to nginx and encrypt all of the things

jyundt opened the pull request:
Pretty significant refactor of the ansible playbook:
* Added Ubuntu 16.04 support
* Upgraded to postgres 9.6
* Migrated from apache -> nginx
* Switched to third party / upstream repos for postgres and nginx
* Included Let's Encrypt integration for all services (all uwsgi/postgres traffic is encrypted with TLSv1.2)
* Created nightly S3 backup for `db` role
* Moved all variables from `ROLENAME/vars` to `group_vars/all/vars.yml` and `group_vars/all/vault.yml`
* Renamed encrypted vars to `vault_VARNAME` as per [Ansible best practices](http://docs.ansible.com/ansible/playbooks_best_practices.html#variables-and-vaults)
* Renewed theoval.us SSL cert and migrated to sha-2

Known issues:
* The current playbook will modify nginx, uwsgi, postgres and firewalld rules to only allow communication between servers identified in the ansible inventory / play.  However, firewalld rules are not "removed" if servers are subsequently removed from the environment.  All services (postgres, nginx, uwsgi) _will_ be properly reconfigured when servers are added/removed from the environment, but firewalld rules are only "added" each time.  This, unfortunately, creates a scenario where multiple "stale" firewalld rules could be present if servers are deleted.

jyundt committed:
Add initial code for ubuntu support.
Modified base role to (almost) work with Ubuntu 16.04 and CentOS 7.

jyundt committed:
checking in for the night

jyundt committed:
checking in for the day.

jyundt committed:
quick check in before heading out.

jyundt committed:
Finished refactor of base and db.
Now supports Ubuntu 16.04 and CentOS 7.

jyundt committed:
Refactor base, db and app roles to support Ubuntu 16.04 and CentOS 7.

jyundt committed:
Fixup to make app and db only listen on a private network.

jyundt committed:
Fix a typo with the oval_user shell.

jyundt committed:
Initial nginx on centos and ubuntu functionality.
TODO: handle theoval.us official ssl cert better.

jyundt committed:
Add certificate (and nginx config) for theoval.us cert.

jyundt committed:
Adding an updated Readme.md

jyundt committed:
Minor fixup for DB migration task for Ubuntu.

jyundt committed:
Workaround strange Ubuntu behavior when trying to perform a db migration/upgrade.

jyundt committed:
Adding nightly S3 backups for the db role.

jyundt committed:
Update SSL certificate for theoval.us

jyundt committed:
Add HSTS for nginx, gotta get that A+ from ssllabs...

jyundt committed:
Update vault with new route53 zone id.

jyundt committed:
Fix misplaced selinux task.

jyundt committed:
Update pg_hba.conf template to include localhost access.

jyundt committed:
Update pg_hba.conf to use localhost for IPv6 and fix a bug with pg_config not in PATH.

jyundt committed:
Update certbot tasks to use the Let's Encrypt staging server for DEV and TST environments.

johnwheffner commented on README.md:
```diff
@@ -3,53 +3,112 @@
+* [app](roles/app) 
+* [web](roles/web) 
 
+Both postgres and nginx are installed from the "upstream" apt/yum repos (instead of the base CentOS or Ubuntu repos).  This provides a semi-consistent across operating systems.
```
provides a semi-consistent... *something?* ... across operating systems

johnwheffner commented on README.md:
````diff
@@ -3,53 +3,112 @@
+web4.example.com ansible_user=root ansible_python_interpreter=/usr/bin/python3
+```
+
+Any servers that use python3 by default (e.g. Ubuntu 16.04) **must** have the `ansible_python_interpreter` variable set to `/usr/bin/python3`.
````
So if the default interpreter is python3, you have to tell it you're using python3?  Is that backwards, or does ansible take behave differently if the python interpreter ends in 3?

jyundt replied:
Ansible (_mostly_) supports python2 and python3.  However, ansible will assume a python interpreter named `python` is available in your `$PATH`. For distributions like Ubuntu 16.04, the default python interpreter is `/usr/bin/python3` (python 2 is **not** installed by default).

So you have two options:

- Log in before Ansible runs and manually install python2
- Set the location of your python interpreter (typically in the inventory file) to `/usr/bin/python3`

johnwheffner commented on README.md:
```diff
@@ -3,53 +3,112 @@
+
+
+### Operating System Specific Variables
+CentOS/Ubuntu specific variables are stored in [group_vars/os_RedHat.yml](group_vars/os_RedHat.yml) and [group_vars/os_Debian](group_vars/os_Debian) respectively.
```
os_Debian needs the .yml suffix

johnwheffner reviewed (commented):
I like the new organization of config files much more.  Thumbs up on that.

johnwheffner commented on roles/base/tasks/main.yml:
```diff
@@ -1,72 +1,222 @@
+    mode: 0600
 
 - name: Swapon swap file
   command: swapon /swapfile
```
Not using encrypted swap??

jyundt replied:
No, but we could if you want to. I know how strongly you feel about encrypting all of the things.

johnwheffner replied:
j/k.  We put all our secrets in the filesystem anyway.

jyundt committed:
Fixup for README.md

jyundt committed:
Update inventory files.

jyundt commented:
@johnwheffner 

I may have done a "bad thing" but I rebased this branch off of master and force pushed to resolve a merge conflict. Sorry about that...

In either case, lemme know if you have any additional questions about the refactored playbook setup.

johnwheffner commented on prd.ini:
```diff
@@ -0,0 +1,13 @@
+[db]
```
Where is this file used?

johnwheffner replied:
Looks like this is your inventory file.  Do we want to have this under version control?

jyundt replied:
Yeah, I think this snuck in by mistake. We can get rid of this inventory file.

johnwheffner commented on roles/db/tasks/main.yml:
```diff
@@ -1,32 +1,165 @@
+    special_time: daily
+    user: postgres
+    state: present
+    job: "/usr/bin/pg_dump {{db_name}}  | xz >  {{pgdata}}/backups/{{db_name}}_db_{{ansible_fqdn}}_$(date +\\%Y\\%m\\%d).sql.xz && aws s3 cp {{pgdata}}/backups/{{db_name}}_db_{{ansible_fqdn}}_$(date +\\%Y\\%m\\%d).sql.xz s3://{{s3_bucket_name}}/{{ansible_fqdn}}/{{db_name}}_db_{{ansible_fqdn}}_$(date +\\%Y\\%m\\%d).sql.xz"
```
I kind of like the idea of periodically running aws s3 sync of the directory rather than aws s3 cp.  That way, if one cp fails, the next sync will copy the failed one and the new one both.

jyundt replied:
Sure that works, I can update the playbook to sync instead of copy.

johnwheffner commented:
I thought you'd added some stuff to automatically update hostnames, etc.  Not part of this change?

jyundt commented:
> I thought you'd added some stuff to automatically update hostnames, etc. Not part of this change?

I had some dynamic inventory generation stuff previously, but it seemed like overkill. I can add it back in if you are interested.

johnwheffner commented:
Just curious, wanted to make sure I wasn't missing something.

johnwheffner commented:
What's the minimum required ansible version to run the playbook?

jyundt commented:
> What's the minimum required ansible version to run the playbook?

I _think_ 2.2, but 2.1 or 2.0 may work.

I did all of my testing with `2.2.1.0`

johnwheffner commented:
Does not work with ansible 2.0, had to upgrade.

johnwheffner commented:
I think I'd rather the ansible user not be root, but rather use 'become' when necessary.

jyundt commented:
> I think I'd rather the ansible user not be root, but rather use 'become' when necessary.

You may get into a chicken-and-egg scenario for the initial ansible play. Some cloud providers (DigitalOcean) only provide an initial root account. In this scenario you'd need to run a separate play to configure a non-root account and setup sudo access.

If you are using something like AWS you can modify your inventory file to include `ansible_user` and set it to `ec2-user` to accomplish the same goal but still keep it relatively provider agnostic.

johnwheffner commented:
Without having the playbook specify become=yes, you can't run as a non-root user.

jyundt commented:
> Without having the playbook specify become=yes, you can't run as a non-root user.

Similar to `ansible_user` and `ansible_python_interpreter`, you should set `become` in the inventory. 

Update: fixing `python_interpreter` -> `ansible_python_interpreter`

johnwheffner commented:
The "single-host.yml" playbook doesn't seem to get the variables imported correctly.  Is this still supposed to work?

jyundt commented:
> The "single-host.yml" playbook doesn't seem to get the variables imported correctly. Is this still supposed to work?

No, that play is deprecated, I should have removed it.

johnwheffner commented:
Next issue: certbot is failing.  Might need some help with that, as I have no idea how that works.  Looks like it's trying to use the hostname that the internal address resolves to.

jyundt commented:
> Next issue: certbot is failing. Might need some help with that, as I have no idea how that works. Looks like it's trying to use the hostname that the internal address resolves to.

Hmm if you are using AWS it could be related to your VPC config or some other DNS weirdness. 

Can you send me logs/detailed information in Slack?

(Or just stop using AWS...)

johnwheffner commented:
Looks like this is ansible, not certbot.  I changed the hostname on the box, and now it's getting a slightly different failure: ansible_fqdn is getting set to ec2-34-205-132-95.compute-1.amazonaws.com, which is the ptr record for 95.132.205.34.in-addr.arpa (my address is 34.205.132.95).  I'll look into that tomorrow.  Time for some sleep.

johnwheffner commented:
Adding an entry in /etc/hosts worked.  Possibly less than ideal.

jyundt committed:
Remove local prd.ini inventory file, update README.md to reflect required Ansible Version and change DB backups from s3 cp to s3 sync.

jyundt committed:
Remove legacy single-host.yml playbook.

johnwheffner commented on roles/base/tasks/main.yml:
```diff
@@ -1,72 +1,222 @@
+    job: "/usr/bin/{{certbot_package}} renew -n --keep --standalone --pre-hook /etc/letsencrypt/hooks/pre.sh --post-hook /etc/letsencrypt/hooks/post.sh"
+  when: oval_environment == "PRD"
+
+- name: Install certbot cronjob (STAGING)
```
The cron job for staging is the same as for prod, right?

jyundt replied:
> The cron job for staging is the same as for prod, right?

I think that's a typo on my part, the STAGING renew cronjob should use `--staging` I _think_.

Lemme verify if `--staging` is needed for the renew process. If it is, I'll fix this.

If it isn't, I'll remove the extra ansible task.

jyundt replied:
Verified that `--staging` is **not** required when renewing, I just updated this task in 162c02f.

jyundt committed:
A few minor fixes:
- Update oval UID/GID to != 1000 (set to 1050)
- Switch from aws s3 copy -> aws s3 sync (for nightly DB backups)
- Remove unnecessary cronjob for PRD vs. STAGING certbot renewal
- Switch from ansible_fqdn -> inventory_hostname for some name resolution stuff

jyundt committed:
Temporary switch from suwsgi -> https-socket until unbit/uwsgi#1490 is fixed.

johnwheffner commented on roles/base/tasks/main.yml:
```diff
@@ -1,72 +1,211 @@
+- name: Check for an active swapfile
+  command: grep "^/swapfile" /proc/swaps
+  register: swapfile_result
   ignore_errors: true
```
Suggest using `changed_when: false`

jyundt replied:
Excellent suggestion, I'll fix that right now.

jyundt committed:
Include changed_when argument for swap file checks in base role.

johnwheffner commented on roles/web/templates/default.conf.j2:
```diff
@@ -0,0 +1,79 @@
+
+server {
+    listen       80;
+    server_name  {{ansible_fqdn}} www.{{ansible_fqdn}};
```
Occurrences of `{{ansible_fqdn}}` here also need to be replaced with `{{inventory_hostname}}`.

jyundt replied:
Good catch, fixed in 8b5fc66.

jyundt committed:
Fixup: replace ansible_fqdn -> inventory_hostname

johnwheffner reviewed (approved):
Much improved, works for me.
Would be good to nail down those uwsgi/nginx errors but let's land this.

jyundt committed:
Fixup: use IP address for connecting to DB instead of FQDN (avoids split DNS reqs).

jyundt committed:
Re-Fixup: reverting back to FQDNs for DB_HOST.
